### PR TITLE
Kernel: Implement support for PS2 scan code set 2

### PR DESCRIPTION
**Pull request by supercomputer7**
Relies on #18574.
Will unblock work on PR #18792 (and also on PR #18673).

### CODE REVIEW

**Comment by supercomputer7** [comment]
> Now that #17945 was merged, I will need to ensure it takes the key index from other value rather the scan code.

Done.

**Comment by supercomputer7** [comment]
There might be other places like it was with `WindowServer` that uses raw scancodes. Blocked until further notice.

**Comment by supercomputer7** [comment]
> There might be other places like it was with `WindowServer` that uses raw scancodes. Blocked until further notice.

Seems like I found all instances now. Still not 100% sure everything works as intended (as I mean "I don't know how to test this changes" for `KeyboardMapper`) but the code seems correct.

File: Kernel/API/KeyCode.h
Comment (BertalanD) [inline Kernel/API/KeyCode.h:176]: It looks like nothing really uses `scancode` in the userland after this PR, so we should rather remove it. That would prevent the potential confusion between between `map_entry_index` and `scancode` (the difference is not obvious to me, at least...)

File: Kernel/Arch/x86_64/ISABus/I8042Controller.h
Comment (BertalanD) [inline Kernel/Arch/x86_64/ISABus/I8042Controller.h:100]: "first port" translation sounds a bit odd to me. If the point of #18654 was to remove assumptions about what's where on the serial bus, this would contradict that. How about we directly query the command line in the `KeyboardDevice` constructor?

File: Kernel/Devices/HID/KeyboardDevice.cpp
Comment (BertalanD) [inline Kernel/Devices/HID/KeyboardDevice.cpp:458]: How much is this Scan Code 1/2 translation logic necessary for the operation of a USB keyboard? If we want `KeyboardDevice` to be a generic device used by both PS/2 and USB, the code would be better factored if this logic was moved into `PS2KeyboardDevice` IMO.

**Review by BertalanD — COMMENTED** [review]
Thank you a thousand times for getting us closer to USB HID support.

I have a few comments about the design, I'm a bit afraid this interface still has some PS2-isms that would be good to avoid.

**Comment by supercomputer7** [comment]
This is blocked until #18574 is merged.

File: Kernel/Arch/x86_64/ISABus/I8042Controller.h
Comment (supercomputer7) [reply Kernel/Arch/x86_64/ISABus/I8042Controller.h:100]: The problem is that PS2 set2->set1 scan code translation can enabled on the first port on the i8042 - this is not something about semantics - it's literally how the i8042 controller works. if you connect the keyboard to the second port, technically everything should work, but you can't have scan codes being translated from set 2 (if the keyboard sends scan code set 2) to set 1. if you connect a mouse to the first port and enable translation, you will get garbage data.

File: Kernel/Devices/HID/KeyboardDevice.cpp
Comment (supercomputer7) [reply Kernel/Devices/HID/KeyboardDevice.cpp:458]: The problem is that there's a strong relation between scan codes to other attributes of the keyboard - like modifiers, which are currently tied quite strongly to given scan codes. I will try refactor it, but for actually getting somewhere in reasonable time, I think we should aim at getting USB keyboard support in and worry about abstractions later on.

File: Kernel/API/KeyCode.h
Comment (supercomputer7) [reply Kernel/API/KeyCode.h:173]: This is not easily removable as the kernel uses this attribute as well. Maybe rename it to "raw_scancode" will be more appropriate and `map_entry_index` => `layout_map_entry_index`?

**Comment by supercomputer7** [comment]
This is ready for review (and possibly merging as well) :)

File: Userland/Services/WindowServer/Window.cpp
Comment (ADKaster) [inline Userland/Services/WindowServer/Window.cpp:512]: Ok, I'm confused about something here.

This "mode" of the PS2 Keyboard. It's static right? After init anyway. We figure out at boot time what "keymap set" is going to be correct for the keyboard device, and then it can't change. We can't hotplug PS2. You can't plug PS2 in after boot. Are any of these assumptions incorrect?

If I have the correct understanding, then why can't the kernel just give us a system character map that represents the actual attached hardware? I don't understand why we need to pass this extra information to userspace in each key event when we could simply present the properly indexed map data to the user via this "fetch system map" call here

File: Kernel/Devices/HID/KeyboardDevice.h
Comment (ADKaster) [inline Kernel/Devices/HID/KeyboardDevice.h:59]: I agree with Daniel, why does this scan code set1/set2 logic need to belong on the base class? It is specific to I8042 devices, and should not be leaked into the higher level abstraction of "a keyboard with unknown physical connection to the computer"

File: Userland/Services/WindowServer/Window.cpp
Comment (supercomputer7) [reply Userland/Services/WindowServer/Window.cpp:512]: What about USB HID then? it's not really useful to hardcode things just to make them flexible later on, especially when USB HID is just around the corner.

Comment (ADKaster) [reply Userland/Services/WindowServer/Window.cpp:512]: What I'm trying to suggest is that we should know, at the time that the user fetches the system character map, which keyboard is active, and what keymap it has. And return that keymap pointer in the `getkeymap` syscall. So I don't see a reason to add this extra complexity to the userspace code when the kernel has all the information.

File: Kernel/Devices/HID/KeyboardDevice.h
Comment (ADKaster) [reply Kernel/Devices/HID/KeyboardDevice.h:59]: @supercomputer7 as far as I see it, this comment is blocking the PR. We should not be adding PS2-specific protected methods to a class that is shared b/w different keyboard devices.

File: Userland/Services/WindowServer/Window.cpp
Comment (ADKaster) [inline Userland/Services/WindowServer/Window.cpp:526]: Do we need this check? it seems that the kernel giving us a Keyboard event with index 255 is completely invalid anyway. should it be an assert or verify instead?

File: Userland/Libraries/LibWeb/UIEvents/KeyboardEvent.cpp
Comment (ADKaster) [inline Userland/Libraries/LibWeb/UIEvents/KeyboardEvent.cpp:473]: This looks overlooked. One of these is surely LeftControl?

**Review by ADKaster — COMMENTED** [review]
Ah, I like this abstraction. it's possible we might be able to re-unify the key maps when USB HID and other keyboard devices are in, but this separation makes a lot more sense looking at the code. I did find one small thing though:

File: Kernel/Devices/HID/PS2/KeyboardDevice.cpp
Comment (ADKaster) [inline Kernel/Devices/HID/PS2/KeyboardDevice.cpp:430]: In a follow-up, can we just drop the keyboard event if it's an unexpected multi-byte sequence, with a complaint in the log? If I wave a magnet over the cable while typing pause-break, we shouldn't crash the kernel.

File: Userland/Services/WindowServer/Window.cpp
Comment (supercomputer7) [reply Userland/Services/WindowServer/Window.cpp:526]: well, index 0xFF could be used for special keys, like Ctrl.

File: Kernel/Devices/HID/PS2/KeyboardDevice.cpp
Comment (supercomputer7) [reply Kernel/Devices/HID/PS2/KeyboardDevice.cpp:430]: Should we do it in this PR? or a future commit?

Comment (ADKaster) [reply Kernel/Devices/HID/PS2/KeyboardDevice.cpp:430]: future

**Review by ADKaster — APPROVED** [review]
Okay, looks good now, thanks for going through with the refactor! Hopefully this is the next step to more HID devices :D

**Comment by ADKaster** [comment]
@supercomputer7 I reverted the last two commits in https://github.com/SerenityOS/serenity/pull/22497 at the main commit of the PR breaks shift and caps lock from using with the default QEMU settings. Please take a look and re-submit the patches when you find the issue :)

**Pull request merged**
